### PR TITLE
More COM release and Shutdown tweaks

### PR DESCRIPTION
**Pull request by MDoerner**
This PR adds disposel of more COM wrappers. In particular, the `VBProjects` collections get disposed.
Since we attach the events to the underlying RCW, now one wrapper gets saved on the `RubberduckParserState` to keep the RCW from being detached.

Admittedly, the changes to make the LINQ queries release everything make things quite a bit less readable. Maybe we could use some `Enumerable` wrapper magic to turn this back mostly into LINQ. (Note that sometimes the element retunred must not be disposed as the only element int the collection.)

In addition to COM wrapper disposal, this PR removes the broadcast of `RUBBERDUCK_SINKING` on addin shutdown and the GC handles in the `SubclassingWindow` and the `DockableToolwindowHost`. Both are obsolete given the remaining shutdown code: everything we do in reaction to the broadcast, we also do on disposal/release and the `SubclassingWindow`s are kept alive by the `DockableToolwindowHost`s, which are kept alive by a static collection in `Windows`, and, in the case of the `RubberduckHooks`, by the IoC container.   

As a final piece, the PR adds COM release logging to the `HostApplicationBase`, which does not inherit from `SafeComWrapper`.

### CODE REVIEW

File: Rubberduck.VBEEditor/Application/HostApplicationBase.cs [hunk 1]
```diff
@@ -68,6 +119,73 @@ public virtual object Run(string name, params object[] args)
             return null;
         }
 
+        private int? _rcwReferenceCount;
```
Comment (bclothier) [inline Rubberduck.VBEEditor/Application/HostApplicationBase.cs]: Seems to me that the `ApplicationHostBase` really should inherit from `SafeComWrapper` so we wouldn't have to copy all that code here?

File: Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProject.cs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.Office.Interop.MSProject;
```
Comment (bclothier) [inline Rubberduck.VBEEditor/SafeComWrappers/VB6/VBProject.cs]: I think this is another rogue reference...

File: Rubberduck.VBEEditor/Application/HostApplicationBase.cs (hunk 1 above)
Comment (MDoerner) [reply Rubberduck.VBEEditor/Application/HostApplicationBase.cs]: I agree. It really is a wrapper for a COM object.
Any concerns @retailcoder ?

File: Rubberduck.VBEEditor/Application/HostApplicationBase.cs [hunk 3]
```diff
@@ -1,54 +1,60 @@
...
             ApplicationName = applicationName;
+        }
+
+        protected HostApplicationBase(IVBE vbe, string applicationName)
+        :base(ApplicationFomVbe(vbe, applicationName))
+        {
+            ApplicationName = applicationName;
+        }
 
+        private static TApplication ApplicationFomComReflection(string applicationName)
```
Comment (bclothier) [inline Rubberduck.VBEEditor/Application/HostApplicationBase.cs]: I believe it's misspelled, missing `r` for the `From`  --- `ApplicationFromComReflection`

**Comment by retailcoder** [comment]
So from this PR onward every call to a SafeComWrapper needs to be enclosed in a `using` scope? The initial PR that brought SafeComWrappers did exactly that, and it was reverted since... we need to document the SafeComWrapper API and make sure they're all disposed everywhere they're used. @Hosch250 can we leverage Roslyn to enforce this?

**Comment by MDoerner** [comment]
Wrapping everything in `using` statements will not do the trick. We have quite a few places where the wrapper is persisted for later use. In this case, someone else has to dispose the objects later.

It will need some restructuring to get the `VBComponent` wrapper under control. In particular, we need to remove it from the QMNs. Instead, we could save a dictionary of them somewhere to provide access to them to whoever really needs them. (Obviously, that would require some cache invalidation work on that dictionary.)

**Comment by WaynePhillipsEA** [comment]
[AddinBridgeLog_withGC.txt](https://github.com/rubberduck-vba/Rubberduck/files/1595063/AddinBridgeLog_withGC.txt)
[AddinBridgeLog_withoutGC.txt](https://github.com/rubberduck-vba/Rubberduck/files/1595065/AddinBridgeLog_withoutGC.txt)
[AddinBridgeLog_withGC_andOleWrappers.txt](https://github.com/rubberduck-vba/Rubberduck/files/1595067/AddinBridgeLog_withGC_andOleWrappers.txt)

New logs attached for the latest pre release.

Logs have been improved;   
    
1) now all LIVE references are tracked for each component, and reported clearly at disconnection of the addin or toolwindow
2) each leaking reference can now be tracked back to its source, as the object pointers now include a wrapper identifier
   For example, if a live reference is reported at disconnection, named 'WRAP000b:$07029e24', just search for where that identifier is first encountered in the log, and by looking at the calls made around that area, you should be able to work out exactly where in the RD source the leak is occurring.

------------------------------
AddinBridgeLog_withGC.txt shows that with the final GC.Collect() call, this is a clean exit for the Rubberduck.Extension, but not the tool windows

AddinBridgeLog_withoutGC.txt shows that without the final GC.Collect() call, there are leaks in:
    _VBProject, _CodePanes, Window, _VBComponent, _CodePane, _VBProjects, IConnectionPointContainer[events]

AddinBridgeLog_withGC_andOleWrappers shows that with the final GC.Collect() plus the new _DockableWindowHost, this is a completely clean exit.

Very good progress!  Well done.

**Comment by WaynePhillipsEA** [comment]
Also IEnumVARIANT leaks, but not consistently.

**Pull request merged**